docs/templates.md

```diff
@@ -0,0 +1,151 @@
+# Templates
+
+Jujutsu supports a functional language to customize output of commands.
+The language consists of literals, keywords, operators, functions, and
+methods.
+
+A couple of `jj` commands accept a template via `-T`/`--template` option.
+[The default templates](../src/config/templates.toml) are defined in
+the `[templates]` section.
+
+## Keywords
+
+### Commit keywords
+
+The following keywords can be used in `jj log`/`jj obslog` templates.
+
+* `description: String`
+* `change_id: ChangeId`
+* `commit_id: CommitId`
+* `author: Signature`
+* `committer: Signature`
+* `working_copies: String`: For multi-workspace repository, indicate
+  working-copy commit as `<workspace name>@`.
+* `current_working_copy: Boolean`: True for the working-copy commit of the
+  current workspace.
+* `branches: String`
+* `tags: String`
+* `git_refs: String`
+* `git_head: String`
+* `divergent: Boolean`: True if the change id corresponds to multiple visible
+  commits.
+* `conflict: Boolean`: True if the commit contains merge conflicts.
+* `empty: Boolean`: True if the commit modifies no files.
+
+### Operation keywords
+
+The following keywords can be used in `jj op log` templates.
+
+* `current_operation: Boolean`
+* `description: String`
+* `id: OperationId`
+* `tags: String`
+* `time: TimestampRange`
+* `user: String`
+
+## Operators
+
+The following operators are supported.
+
+* `x.f()`: Method call.
+* `x ++ y`: Concatenate `x` and `y` templates.
+
+## Global functions
+
+The following functions are defined.
+
+* `label(label: Template, content: Template) -> Template`: Apply label to
+  the content. The `label` is evaluated as a space-separated string.
+* `if(condition: Boolean, then: Template[, else: Template]) -> Template`:
+  Conditionally evaluate `then`/`else` template content.
+* `concat(content: Template...) -> Template`:
+  Same as `content_1 ++ ... ++ content_n`.
+* `separate(separator: Template, content: Template...) -> Template`:
+  Insert separator between non-empty contents.
```

Comment:
I was just surprised by this in https://github.com/martinvonz/jj/pull/1328. Why do we ignore empty ones?

Reply:
I figured this out. This is so that `separate(" ", id, conflicts, branches, date)` doesn't insert two spaces if there are neither conflicts nor branches.